Source/SwiftLintFramework/Rules/Style/SwitchCaseSortRule.swift

```diff
@@ -0,0 +1,243 @@
+    } else if let stringLiteralExpression = expressionPattern.expression
+        .as(StringLiteralExprSyntax.self) {
+        let segments = stringLiteralExpression.segments.map { segment in
+            switch segment {
+            // ignore interpolation and join the string literals
+            case let .stringSegment(segment):
+                return segment.content.text.removingCommonLeadingWhitespaceFromLines()
+            case .expressionSegment: // string interpolation part
+                return ""
+            }
+        }.joined()
+        return segments.isEmpty ? nil : segments
+    } else if let integerExpression = expressionPattern.expression
+        .as(IntegerLiteralExprSyntax.self) {
+        return integerExpression.digits.text
+    } else if let floatExpression = expressionPattern.expression
+        .as(FloatLiteralExprSyntax.self) {
+        return floatExpression.floatingDigits.text
+    } else if let identifierExpression = expressionPattern.expression
+        .as(IdentifierExprSyntax.self) {
+        return identifierExpression.identifier.text
+    } else if let functionCall = expressionPattern.expression.as(FunctionCallExprSyntax.self),
+              let memberAccess = functionCall.calledExpression.as(MemberAccessExprSyntax.self) {
+        return memberAccess.name.text
+    }
+    return nil
+}
+
+private func sortableName(for caseItemSyntax: CaseItemSyntax) -> String? {
+    if let expressionPattern = caseItemSyntax.pattern.as(ExpressionPatternSyntax.self) {
+        return sortableName(for: expressionPattern)
+    } else if let valueExpression = caseItemSyntax.pattern.as(ValueBindingPatternSyntax.self)?.valuePattern
+        .as(ExpressionPatternSyntax.self) {
+        return sortableName(for: valueExpression)
+    }
+    return nil
+}
+
+private func sortableName(for caseLabelSyntax: SwitchCaseLabelSyntax) -> String? {
+    if let item = caseLabelSyntax.caseItems.first {
+        return sortableName(for: item)
+    }
+    return nil
+}
+
+private func byName(_ lhs: CaseItemSyntax, _ rhs: CaseItemSyntax) -> Bool {
+    guard let lhsName = sortableName(for: lhs),
+          let rhsName = sortableName(for: rhs) else {
+        return false
+    }
+    return lhsName < rhsName
+}
+
+private func byName(_ lhs: SwitchCaseLabelSyntax, _ rhs: SwitchCaseLabelSyntax) -> Bool {
+    guard let lhsName = sortableName(for: lhs),
+          let rhsName = sortableName(for: rhs) else {
+        return false
+    }
+    return lhsName < rhsName
+}
+
+private func byName(_ lhs: SwitchCaseSyntax, _ rhs: SwitchCaseSyntax) -> Bool {
+    switch (lhs.label, rhs.label) {
+    // default is always last
+    case (.default, _):
+        return false
+    case (_, .default):
+        return true
+    case let (.case(lhsCaseLabel), .case(rhsCaseLabel)):
+        return byName(lhsCaseLabel, rhsCaseLabel)
+    }
+}
+
+private func byName(_ lhs: SwitchCaseListSyntax.Element, _ rhs: SwitchCaseListSyntax.Element) -> Bool {
```

Comment:
What about extensions for these types that implement `Comparable` instead of these freestanding functions used for sorting?

The same extensions could contain the `sortableName` methods/accessors.

`sortableName` could also be extracted by another `SyntaxTransformVisitor` that would then contain all this logic.

Freestanding functions appear without context. They look like arbitrary functions but actually here a specific extraction logic is implemented for all these types. So the functions all belong together. But the don't as each of them stands independent of the others.